src/credentials/LastKnownGoodTime.cpp

```diff
@@ -0,0 +1,251 @@
+/*
+ *
+ *    Copyright (c) 2021-2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @brief Defines a table of fabrics that have provisioned the device.
+ */
+
+#include "LastKnownGoodTime.h"
+
+#include <lib/support/BufferWriter.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+#include <platform/ConfigurationManager.h>
+//#include <lib/core/CHIPSafeCasts.h>
+#include <lib/support/SafeInt.h>
+
+namespace chip {
+
+namespace {
+// Tags for Last Known Good Time.
+constexpr TLV::Tag kLastKnownGoodChipEpochSecondsTag         = TLV::ContextTag(0);
+constexpr TLV::Tag kFailSafeLastKnownGoodChipEpochSecondsTag = TLV::ContextTag(1);
+} // anonymous namespace
+
+void LastKnownGoodTime::LogTime(const char * msg, System::Clock::Seconds32 chipEpochTime)
+{
+    char buf[26] = { 0 }; // strlen("00000-000-000T000:000:000") == 25
+    uint16_t year;
+    uint8_t month;
+    uint8_t day;
+    uint8_t hour;
+    uint8_t minute;
+    uint8_t second;
+    ChipEpochToCalendarTime(chipEpochTime.count(), year, month, day, hour, minute, second);
+    snprintf(buf, sizeof(buf), "%04u-%02u-%02uT%02u:%02u:%02u", year, month, day, hour, minute, second);
+    ChipLogProgress(TimeService, "%s%s", msg, buf);
+}
+
+CHIP_ERROR LastKnownGoodTime::LoadLastKnownGoodChipEpochTime(System::Clock::Seconds32 & lastKnownGoodChipEpochTime,
+                                                             Optional<System::Clock::Seconds32> & failSafeBackup) const
+{
+    TLV::ContiguousBufferTLVReader reader;
+    uint8_t buf[LastKnownGoodTimeTLVMaxSize()];
+    uint16_t size = sizeof(buf);
+    uint32_t seconds;
+    reader.Init(buf, size);
```

Comment:
This needs to happen after we know the actual size of the initialized data, no?  as in, after we get the value from storage.

Reply:
Ah, you are correct.  I missed that size is an in/out pararmeter in the storage getter.

Reply:
Fixed.